File: frontend/src/api/es.ts
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query"
import { apiFetch } from "./client"
import { useClusterContext } from "@/context/ClusterContext"

/**
 * The single client-side auto-refresh cadence (ms) every snapshot hook re-reads the DB cache on,
 * chosen via the TopBar dropdown. Returns `false` when the user picked "Off" (freeze — no polling).
 * Note: this only re-reads the cheap cached snapshot; it never re-polls the live cluster. Only the
 * explicit force-refresh button (`useRefreshSnapshots`) hits Elasticsearch.
 */
function useAutoRefetch(): number | false {
  const { refreshInterval } = useClusterContext()
  return refreshInterval > 0 ? refreshInterval : false
}

export interface ClusterHealth {
  cluster_name: string
  status: string
  number_of_nodes: number
  number_of_data_nodes: number
  active_primary_shards: number
  active_shards: number
  relocating_shards: number
  initializing_shards: number
  unassigned_shards: number
}

export interface NodeInfo {
  name: string
  role: string
  ip: string
  version: string
  disk_total: number
  disk_used: number
  disk_used_percent: number
  heap_max: number
  heap_current: number
  heap_percent: number
  cpu: number
  load_1m: number
  segments_count: number
}

export interface IndexInfo {
  health: string
  status: string
  index: string
  pri: number
  rep: number
  docs_count: number
  store_size: number
  pri_store_size: number
}

export interface ShardInfo {
  index: string
  shard: number
  prirep: string
  state: string
  docs: number
  store: number
  node: string | null
  segments_count: number
}

export interface RecoveryInfo {
  index: string
  shard: number
  source_node: string
  target_node: string
  bytes_total: number
  bytes_recovered: number
  bytes_percent: string
}

export interface StorageGroup {
  name: string
  size_bytes: number
}

export interface NodeRoleCounts {
  master: number
  data: number
  coord: number
  ingest: number
  other: number
}

export interface OverviewData {
  health: ClusterHealth
  index_count: number
  nodes: NodeInfo[]
  recoveries: RecoveryInfo[]
  storage_breakdown: StorageGroup[]
  node_role_counts: NodeRoleCounts
}

/**
 * Every `/es/*` read endpoint now wraps its payload in this envelope so the UI can show staleness.
 * `fetched_at` is a naive UTC ISO string (no `Z` suffix) — parse it as UTC, see `parseUtc`.
 */
export interface Cached<T> {
  data: T
  fetched_at: string
  stale_seconds: number
  next_poll_in: number | null
}

/** Snapshot freshness metadata, surfaced to the TopBar staleness indicator. */
export interface SnapshotMeta {
  fetched_at: string
  stale_seconds: number
  next_poll_in: number | null
}

/** A node row precomputed server-side with its shard count and tier label. */
export interface NodeInfoEx extends NodeInfo {
  shard_count: number
  tier: string
}

/** One shard chip in a grid cell (the minimal payload the chip needs). */
export interface ShardCell {
  shard: number
  prirep: string
  state: string
  store: number
  docs: number
  segments_count: number
}

export interface ShardMapColumn {
  name: string
  short: string
  tier: string
  disk_used_percent: number
}

export interface ShardMapRow {
  index: string
  pri_store_size: number
  health: string
}

/** Precomputed grid: columns = `data_nodes`, rows = `indices`, cell = `cells["<index> <node>"]`. */
export interface ShardMapGrid {
  data_nodes: ShardMapColumn[]
  indices: ShardMapRow[]
  cells: Record<string, ShardCell[]>
}

/** Per-data-node aggregate at a pivot tree node. */
export interface PivotNodeAgg {
  node: string
  shard_count: number
  size: number
}

/** A node in the dynamic-depth pivot rollup tree. */
export interface PivotNode {
  key: string
  label: string
  depth: number
  total_size: number
  total_docs: number
  shard_count: number
  index_count: number
  per_node: PivotNodeAgg[]
  children: PivotNode[]
  is_leaf: boolean
}

export interface PivotColumn {
  name: string
  short: string
  disk_used_percent: number
}

/** Precomputed dynamic-depth pivot tree with per-data-node aggregates. */
export interface PivotTree {
  separator: string
  data_nodes: PivotColumn[]
  max_cell_size: number
  roots: PivotNode[]
}

export interface Opportunity {
  type: string
  severity: string
  detail: string
  wasted_shards: number
  target_shards: number
}

export interface AnalyzedIndex {
  name: string
  health: string
  status: string
  pri_count: number
  rep_count: number
  doc_count: number
  pri_store_bytes: number
  store_bytes: number
  avg_shard_size_gb: number
  max_shard_size_gb: number
  max_segments_per_shard: number
  shard_size_cv: number
  opportunities: Opportunity[]
  opportunity_count: number
  wasted_shards: number
}

export interface AnalysisData {
  total_indices: number
  total_with_opportunities: number
  total_wasted_shards: number
  indices: AnalyzedIndex[]
}

function esPath(clusterId: number, path: string) {
  return `/api/clusters/${clusterId}/es${path}`
}

/**
 * Parse the snapshot `fetched_at` string. The backend serializes a naive UTC datetime (no `Z`), so
 * we append `Z` when no timezone designator is present to avoid local-time misinterpretation.
 */
export function parseUtc(iso: string): number {
  const hasTz = /[zZ]$|[+-]\d{2}:?\d{2}$/.test(iso)
  return new Date(hasTz ? iso : `${iso}Z`).getTime()
}

/** Extract just the freshness metadata from a wrapped response. */
function pickMeta<T>(r: Cached<T>): SnapshotMeta {
  return { fetched_at: r.fetched_at, stale_seconds: r.stale_seconds, next_poll_in: r.next_poll_in }
}

export function useClusterHealth(clusterId: number | null) {
  const refetchInterval = useAutoRefetch()
  return useQuery({
    queryKey: ["es", "health", clusterId],
    queryFn: () => apiFetch<Cached<ClusterHealth>>(esPath(clusterId!, "/health")),
    enabled: clusterId !== null,
    refetchInterval,
    select: (r) => r.data,
  })
}

export function useOverview(clusterId: number | null) {
  const refetchInterval = useAutoRefetch()
  return useQuery({
    queryKey: ["es", "overview", clusterId],
    queryFn: () => apiFetch<Cached<OverviewData>>(esPath(clusterId!, "/overview")),
    enabled: clusterId !== null,
    refetchInterval,
    select: (r) => r.data,
  })
}

export function useNodes(clusterId: number | null) {
  const refetchInterval = useAutoRefetch()
  return useQuery({
    queryKey: ["es", "nodes", clusterId],
    queryFn: () => apiFetch<Cached<NodeInfoEx[]>>(esPath(clusterId!, "/nodes")),
    enabled: clusterId !== null,
    refetchInterval,
    select: (r) => r.data,
  })
}

export function useShards(clusterId: number | null) {
  const refetchInterval = useAutoRefetch()
  return useQuery({
    queryKey: ["es", "shards", clusterId],
    queryFn: () => apiFetch<Cached<ShardInfo[]>>(esPath(clusterId!, "/shards")),
    enabled: clusterId !== null,
    refetchInterval,
    select: (r) => r.data,
  })
}

export function useShardMap(clusterId: number | null) {
  const refetchInterval = useAutoRefetch()
  return useQuery({
    queryKey: ["es", "shard-map", clusterId],
    queryFn: () => apiFetch<Cached<ShardMapGrid>>(esPath(clusterId!, "/shard-map")),
    enabled: clusterId !== null,
    refetchInterval,
    select: (r) => r.data,
  })
}

export function usePivot(clusterId: number | null) {
  const refetchInterval = useAutoRefetch()
  return useQuery({
    queryKey: ["es", "pivot", clusterId],
    queryFn: () => apiFetch<Cached<PivotTree>>(esPath(clusterId!, "/pivot")),
    enabled: clusterId !== null,
    refetchInterval,
    select: (r) => r.data,
  })
}

export function useAnalysis(clusterId: number | null, problemsOnly: boolean = false) {
  const refetchInterval = useAutoRefetch()
  return useQuery({
    queryKey: ["es", "analyze", clusterId, problemsOnly],
    queryFn: () =>
      apiFetch<Cached<AnalysisData>>(
        esPath(clusterId!, `/analyze${problemsOnly ? "?problems_only=true" : ""}`),
      ),
    enabled: clusterId !== null,
    refetchInterval,
    staleTime: 60_000,
    select: (r) => r.data,
  })
}

/**
 * Surface snapshot freshness for the TopBar staleness indicator. Reads the lightweight `health`
 * snapshot on the single client auto-refresh cadence; the query's `dataUpdatedAt` drives the
 * TopBar "next in Ns" countdown so it always matches the chosen interval.
 */
export function useSnapshotMeta(clusterId: number | null) {
  const refetchInterval = useAutoRefetch()
  return useQuery({
    queryKey: ["es", "health", clusterId],
    queryFn: () => apiFetch<Cached<ClusterHealth>>(esPath(clusterId!, "/health")),
    enabled: clusterId !== null,
    refetchInterval,
    select: pickMeta,
  })
}

/**
 * Force a server-side snapshot refresh (`POST /es/refresh`). The backend rebuilds every kind from
 * one raw ES fetch, so we invalidate all `es` queries on success to pull the fresh snapshots.
 */
export function useRefreshSnapshots(clusterId: number | null) {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: () =>
      apiFetch<{ cluster_id: number; kind: string | null; fetched_at: string | null }>(
        esPath(clusterId!, "/refresh"),
        { method: "POST" },
      ),
    onSuccess: () => qc.invalidateQueries({ queryKey: ["es"] }),
  })
}

export interface ClusterSettings {
  persistent: Record<string, string>
  transient: Record<string, string>
  defaults: Record<string, string>
}

export function useClusterSettings(clusterId: number | null) {
  return useQuery({
    queryKey: ["es", "settings", clusterId],
    queryFn: () => apiFetch<ClusterSettings>(esPath(clusterId!, "/settings")),
    enabled: clusterId !== null,
  })
}

export function useUpdateSettings(clusterId: number | null) {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (body: { persistent?: Record<string, string | null>; transient?: Record<string, string | null> }) =>
      apiFetch(esPath(clusterId!, "/settings"), { method: "PUT", body: JSON.stringify(body) }),
    onSuccess: () => qc.invalidateQueries({ queryKey: ["es", "settings"] }),
  })
}

export interface RebalanceSuggestion {
  index: string
  shard: number
  from_node: string
  to_node: string
  size_bytes: number
}

export function useRebalanceSuggestions(clusterId: number | null) {
  return useQuery({
    queryKey: ["es", "rebalance-suggestions", clusterId],
    queryFn: () =>
      apiFetch<{ suggestions: RebalanceSuggestion[] }>(esPath(clusterId!, "/rebalance-suggestions")),
    enabled: clusterId !== null,
  })
}

File: backend/routes/es.py
from collections.abc import Awaitable, Callable
from datetime import UTC, datetime
from typing import Any

from fastapi import APIRouter, Depends
from pydantic import BaseModel
from sqlalchemy.ext.asyncio import AsyncSession

from backend.auth import require_cluster_access
from backend.database import get_db
from backend.dependencies import get_es_client, require_writable_cluster
from backend.schemas.analysis import AnalysisResponse
from backend.schemas.es import (
    CachedResponse,
    ClusterHealthResponse,
    IndexInfo,
    NodeInfo,
    NodeInfoEx,
    OverviewResponse,
    RecoveryInfo,
    ShardInfo,
    StorageGroup,
)
from backend.services import snapshot_repo
from backend.services.es_client import ESClient
from config.settings import PollingSettings, get_settings

router = APIRouter(
    prefix="/api/clusters/{cluster_id}/es",
    tags=["elasticsearch"],
    dependencies=[Depends(require_cluster_access)],
)

# Which polling cadence governs each snapshot kind, so reads can report a truthful ``next_poll_in``.
# The poller has exactly two ticks: the light tick refreshes ONLY ``health`` (``health_seconds``);
# every other kind (overview, nodes, indices, shards, shardmap, pivot) is rebuilt on the heavy tick
# (``heavy_seconds``), since they all depend on the heavy ``_cat/shards`` + ``_cat/nodes`` fetch.
_LIGHT_KINDS = frozenset({"health"})


def _safe_int(val, default: int = 0) -> int:
    if val is None or val == "":
        return default
    try:
        return int(val)
    except (ValueError, TypeError):
        return default


def _safe_float(val, default: float = 0.0) -> float:
    if val is None or val == "":
        return default
    try:
        return float(val)
    except (ValueError, TypeError):
        return default


def _parse_node(raw: dict) -> NodeInfo:
    role = raw.get("node.role", "")
    if not role or role == "-":
        role = "coord"
    return NodeInfo(
        name=raw.get("name", ""),
        role=role,
        ip=raw.get("ip", ""),
        version=raw.get("version", ""),
        disk_total=_safe_int(raw.get("disk.total")),
        disk_used=_safe_int(raw.get("disk.used")),
        disk_used_percent=_safe_float(raw.get("disk.used_percent")),
        heap_max=_safe_int(raw.get("heap.max")),
        heap_current=_safe_int(raw.get("heap.current")),
        heap_percent=_safe_float(raw.get("heap.percent")),
        cpu=_safe_float(raw.get("cpu")),
        load_1m=_safe_float(raw.get("load_1m")),
        segments_count=_safe_int(raw.get("segments.count")),
    )


def _parse_index(raw: dict) -> IndexInfo:
    return IndexInfo(
        health=raw.get("health", ""),
        status=raw.get("status", ""),
        index=raw.get("index", ""),
        pri=_safe_int(raw.get("pri")),
        rep=_safe_int(raw.get("rep")),
        docs_count=_safe_int(raw.get("docs.count")),
        docs_deleted=_safe_int(raw.get("docs.deleted")),
        store_size=_safe_int(raw.get("store.size")),
        pri_store_size=_safe_int(raw.get("pri.store.size")),
    )


def _parse_shard(raw: dict) -> ShardInfo:
    return ShardInfo(
        index=raw.get("index", ""),
        shard=_safe_int(raw.get("shard")),
        prirep=raw.get("prirep", ""),
        state=raw.get("state", ""),
        docs=_safe_int(raw.get("docs")),
        store=_safe_int(raw.get("store")),
        node=raw.get("node"),
        segments_count=_safe_int(raw.get("segments.count")),
    )


def _parse_recovery(raw: dict) -> RecoveryInfo:
    return RecoveryInfo(
        index=raw.get("index", ""),
        shard=_safe_int(raw.get("shard")),
        source_node=raw.get("source_node", ""),
        target_node=raw.get("target_node", ""),
        bytes_total=_safe_int(raw.get("bytes_total")),
        bytes_recovered=_safe_int(raw.get("bytes_recovered")),
        bytes_percent=raw.get("bytes_percent", "0%"),
    )


def _build_storage_breakdown(raw_indices: list[dict], max_groups: int = 10) -> list[StorageGroup]:
    groups: dict[str, int] = {}
    for raw in raw_indices:
        name = raw.get("index", "")
        prefix = name.split("-")[0].split("_")[0] if name else "unknown"
        prefix = prefix.lstrip(".")
        if not prefix:
            prefix = "unknown"
        groups[prefix] = groups.get(prefix, 0) + _safe_int(raw.get("store.size"))
    sorted_groups = sorted(groups.items(), key=lambda x: x[1], reverse=True)
    top = sorted_groups[:max_groups]
    rest_bytes = sum(v for _, v in sorted_groups[max_groups:])
    result = [StorageGroup(name=name, size_bytes=size) for name, size in top]
    if rest_bytes > 0:
        result.append(StorageGroup(name="other", size_bytes=rest_bytes))
    return result


def _now() -> datetime:
    """Naive UTC ``now`` matching the naive ``fetched_at`` the snapshot repo stores."""
    return datetime.now(UTC).replace(tzinfo=None)


def _poll_interval(kind: str, polling: PollingSettings) -> int:
    """Seconds the poller waits between refreshes of ``kind`` (drives ``next_poll_in``)."""
    if kind in _LIGHT_KINDS:
        return polling.health_seconds
    return polling.heavy_seconds


async def _serve(
    db: AsyncSession,
    cluster_id: int,
    kind: str,
    build_live: Callable[[], Awaitable[Any]],
) -> CachedResponse:
    """Snapshot-first read with live fallback, wrapped in :class:`CachedResponse`.

    If a snapshot row exists for ``(cluster_id, kind)``, serve it with a computed staleness. If not
    (the poller hasn't run yet, or polling is disabled), run ``build_live`` to compute today's live
    payload, best-effort upsert it so the next read is cached, and return it with ``stale_seconds==0``.
    Never returns empty when the cluster is reachable.
    """
    polling = get_settings().polling
    interval = _poll_interval(kind, polling)

    snapshot = await snapshot_repo.get_latest(db, cluster_id, kind)
    if snapshot is not None:
        stale_seconds = max(0, int((_now() - snapshot.fetched_at).total_seconds()))
        return CachedResponse(
            data=snapshot.payload,
            fetched_at=snapshot.fetched_at,
            stale_seconds=stale_seconds,
            next_poll_in=max(0, interval - stale_seconds),
        )

    payload = await build_live()
    item_count = len(payload) if isinstance(payload, list) else payload.get("index_count", 0)
    fetched = await snapshot_repo.upsert_snapshot(db, cluster_id, kind, payload, item_count, 0)
    return CachedResponse(
        data=payload,
        fetched_at=fetched.fetched_at,
        stale_seconds=0,
        next_poll_in=interval,
    )


# --- Live builders (the fallback path; reuse the same pure builders the poller uses) -------------
#
# ``snapshot_service`` imports the parsers from this module, so it is imported lazily here to avoid a
# circular import at module load. It is fully initialized by the time any request runs these.


async def _live_health(es: ESClient) -> dict:
    return await es.cluster_health()


async def _live_overview(es: ESClient) -> dict:
    from backend.services import snapshot_service

    health = await es.cluster_health()
    raw_nodes = await es.cat_nodes_detailed()
    raw_indices = await es.cat_indices_detailed()
    raw_recoveries = await es.cat_recovery_active()
    nodes = [_parse_node(n).model_dump() for n in raw_nodes]
    indices = [_parse_index(i).model_dump() for i in raw_indices]
    recoveries = [_parse_recovery(r).model_dump() for r in raw_recoveries]
    return snapshot_service.build_overview(health, nodes, indices, recoveries, raw_indices)


async def _live_nodes(es: ESClient) -> list[dict]:
    from backend.services import snapshot_service

    raw_nodes = await es.cat_nodes_detailed()
    raw_shards = await es.cat_shards_detailed()
    nodes = [_parse_node(n).model_dump() for n in raw_nodes]
    shards = [_parse_shard(s).model_dump() for s in raw_shards]
    return snapshot_service.build_nodes(nodes, shards)


async def _live_indices(es: ESClient) -> dict:
    from backend.services import snapshot_service

    raw_indices = await es.cat_indices_detailed()
    raw_shards = await es.cat_shards_detailed()
    indices = [_parse_index(i).model_dump() for i in raw_indices]
    shards = [_parse_shard(s).model_dump() for s in raw_shards]
    return snapshot_service.build_indices(indices, shards)


async def _live_shards(es: ESClient) -> list[dict]:
    from backend.services import snapshot_service

    raw_shards = await es.cat_shards_detailed()
    shards = [_parse_shard(s).model_dump() for s in raw_shards]
    return snapshot_service.build_shards(shards)


async def _live_shardmap(es: ESClient) -> dict:
    from backend.services import snapshot_service

    raw_nodes = await es.cat_nodes_detailed()
    raw_indices = await es.cat_indices_detailed()
    raw_shards = await es.cat_shards_detailed()
    nodes = [_parse_node(n).model_dump() for n in raw_nodes]
    indices = [_parse_index(i).model_dump() for i in raw_indices]
    shards = [_parse_shard(s).model_dump() for s in raw_shards]
    return snapshot_service.build_shardmap(nodes, indices, shards)


async def _live_pivot(es: ESClient) -> dict:
    from backend.services import snapshot_service
    from config.settings import get_settings

    raw_nodes = await es.cat_nodes_detailed()
    raw_indices = await es.cat_indices_detailed()
    raw_shards = await es.cat_shards_detailed()
    nodes = [_parse_node(n).model_dump() for n in raw_nodes]
    indices = [_parse_index(i).model_dump() for i in raw_indices]
    shards = [_parse_shard(s).model_dump() for s in raw_shards]
    return snapshot_service.build_pivot(indices, shards, nodes, sep=get_settings().polling.pivot_separator)


# --- Read endpoints (snapshot-first, live fallback) ---------------------------------------------


@router.get("/health", response_model=CachedResponse[ClusterHealthResponse])
async def cluster_health(
    cluster_id: int,
    db: AsyncSession = Depends(get_db),
    es: ESClient = Depends(get_es_client),
):
    return await _serve(db, cluster_id, "health", lambda: _live_health(es))


@router.get("/nodes", response_model=CachedResponse[list[NodeInfoEx]])
async def list_nodes(
    cluster_id: int,
    db: AsyncSession = Depends(get_db),
    es: ESClient = Depends(get_es_client),
):
    return await _serve(db, cluster_id, "nodes", lambda: _live_nodes(es))


@router.get("/indices", response_model=CachedResponse[AnalysisResponse])
async def list_indices(
    cluster_id: int,
    problems_only: bool = False,
    db: AsyncSession = Depends(get_db),
    es: ESClient = Depends(get_es_client),
):
    served = await _serve(db, cluster_id, "indices", lambda: _live_indices(es))
    return _apply_problems_only(served, problems_only)


@router.get("/shards", response_model=CachedResponse[list[ShardInfo]])
async def list_shards(
    cluster_id: int,
    db: AsyncSession = Depends(get_db),
    es: ESClient = Depends(get_es_client),
):
    return await _serve(db, cluster_id, "shards", lambda: _live_shards(es))


@router.get("/overview", response_model=CachedResponse[OverviewResponse])
async def overview(
    cluster_id: int,
    db: AsyncSession = Depends(get_db),
    es: ESClient = Depends(get_es_client),
):
    return await _serve(db, cluster_id, "overview", lambda: _live_overview(es))


@router.get("/shard-map", response_model=CachedResponse[dict])
async def shard_map(
    cluster_id: int,
    db: AsyncSession = Depends(get_db),
    es: ESClient = Depends(get_es_client),
):
    return await _serve(db, cluster_id, "shardmap", lambda: _live_shardmap(es))


@router.get("/pivot", response_model=CachedResponse[dict])
async def pivot(
    cluster_id: int,
    db: AsyncSession = Depends(get_db),
    es: ESClient = Depends(get_es_client),
):
    return await _serve(db, cluster_id, "pivot", lambda: _live_pivot(es))


def _apply_problems_only(served: CachedResponse, problems_only: bool) -> CachedResponse:
    """Filter the served ``indices`` payload to opportunity-bearing indices, server-side.

    The poller stores the full analyzed list once; ``problems_only`` is a trivial read-time filter so
    no second payload is stored. Recomputes the totals over the filtered list to stay consistent.
    """
    if not problems_only:
        return served
    payload = served.data
    filtered = [i for i in payload["indices"] if i["opportunity_count"] > 0]
    served.data = {
        "total_indices": len(filtered),
        "total_with_opportunities": len(filtered),
        "total_wasted_shards": sum(i["wasted_shards"] for i in filtered),
        "indices": filtered,
    }
    return served


@router.get("/analyze", response_model=CachedResponse[AnalysisResponse])
async def analyze_indices(
    cluster_id: int,
    problems_only: bool = False,
    db: AsyncSession = Depends(get_db),
    es: ESClient = Depends(get_es_client),
):
    served = await _serve(db, cluster_id, "indices", lambda: _live_indices(es))
    return _apply_problems_only(served, problems_only)


@router.post("/refresh")
async def refresh(
    cluster_id: int,
    kind: str | None = None,
    db: AsyncSession = Depends(get_db),
    es: ESClient = Depends(get_es_client),
):
    """Force a snapshot refresh on demand. Read-only against ES.

    Runs the full :func:`refresh_cluster` (fetches raw ES once and rebuilds every kind), then reports
    the fresh ``fetched_at`` for the requested ``kind`` (or ``health`` when none is given). ``kind`` is
    a hint for which snapshot's timestamp to return; the underlying refresh always rebuilds all kinds
    from the single raw fetch (so a one-off refresh never costs extra ES round-trips).
    """
    from backend.services import snapshot_service

    await snapshot_service.refresh_cluster(es, cluster_id, db, sep=get_settings().polling.pivot_separator)

    snapshot = await snapshot_repo.get_latest(db, cluster_id, kind or "health")
    return {
        "cluster_id": cluster_id,
        "kind": kind,
        "fetched_at": snapshot.fetched_at if snapshot is not None else None,
    }


@router.get("/rebalance-suggestions")
async def rebalance_suggestions(
    cluster_id: int,  # noqa: ARG001 — required path param for router prefix resolution
    es: ESClient = Depends(get_es_client),
):
    """Advisory: suggest relocate moves to even shard count across data nodes. Read-only against ES."""
    from backend.services import rebalance

    nodes = await es.cat_nodes_detailed()
    shards = await es.cat_shards_detailed()
    return {"suggestions": rebalance.suggest_moves(nodes, shards)}


@router.get("/settings")
async def get_settings_endpoint(es: ESClient = Depends(get_es_client)):
    return await es.cluster_settings_full()


class SettingsUpdateRequest(BaseModel):
    persistent: dict | None = None
    transient: dict | None = None


@router.put("/settings")
async def update_settings(
    cluster_id: int,
    body: SettingsUpdateRequest,
    db: AsyncSession = Depends(get_db),
    es: ESClient = Depends(get_es_client),
):
    """Update cluster settings. Blocked on read-only clusters (this is a cluster write)."""
    await require_writable_cluster(cluster_id, db)
    payload: dict = {}
    if body.persistent is not None:
        payload["persistent"] = body.persistent
    if body.transient is not None:
        payload["transient"] = body.transient
    return await es.put_cluster_settings(payload)


class RestRequest(BaseModel):
    method: str
    path: str
    body: dict | None = None


# REST-console verbs that only read; everything else is treated as a write on a read-only cluster.
_REST_READ_METHODS = frozenset({"GET", "HEAD", "OPTIONS"})


@router.post("/rest")
async def rest_proxy(
    cluster_id: int,
    req: RestRequest,
    db: AsyncSession = Depends(get_db),
    es: ESClient = Depends(get_es_client),
):
    """Proxy an arbitrary ES request. On read-only clusters, only read verbs (GET/HEAD) are allowed.

    The FastAPI request method is always POST here; the verb that matters is ``req.method`` — the ES
    verb carried in the request body. Block any non-read verb on a read-only cluster.
    """
    if req.method.upper() not in _REST_READ_METHODS:
        await require_writable_cluster(cluster_id, db)
    return await es.proxy(req.method, req.path, req.body)
